lib/domain/libBlockchain/BlockchainFactory.h

```
@@ -12,18 +12,16 @@
 #include "BlockchainModel.h"
 
 namespace vanguard {
-    class BlockchainFactory : public LLVMFactory {
+    class BlockchainFactory : public UnitFactory {
     public:
-        static UnitFactory *getInstance() {
+        static BlockchainFactory *getInstance() {
             if(instance == nullptr) {
-                instance = new LLVMFactory();
+                instance = new BlockchainFactory();
             }
 
-            return instance;
+            return dynamic_cast<BlockchainFactory *>(instance);
```

The dynamic cast here is unnecessary.

---

The dynamic cast is necessary because the type of instance is UnitFactory and here we are returning a BlockchainFactory

---

But we know locally that instance is of type BlockChainFactory, you can simply do `(BlockchainFactory *) instance`, which avoids the runtime check.

---

If someone interacts with the factories incorrectly (i.e. both BlockchainFactory and LLVMFactory getInstance are called) that could lead to errors that are more difficult to diagnose than accessing null.

---

Reviewing the API again, it seems there is a need for a dynamic cast if we want to throw an exception when clients call the wrong getInstance method. If this is the case, maybe it's better to remove the getInstace from all subclasses of UnitFactory and replace them with an init function.

---

@stephensj2: I saw your comment after I wrote mine. Yes, I agree with that. Do you want to refactor the API or leave it as is?